pytorch_lightning/loops/optimizer/optimizer_loop.py

```diff
@@ -65,7 +65,7 @@ def connect(self, **kwargs: "Loop") -> None:
         raise NotImplementedError(f"{self.__class__.__name__} does not connect any child loops.")
 
     def reset(self) -> None:
-        if not self.restarting:
+        if not self.restarting or self.done:
```

Comment:
on a larger scope, I will work on a test for restarting in a failed optimizer loop in a separate PR. It's on the todo list #9130

Reply:
Do you think that everywhere we check `not self.restarting` we should also add `or self.done`?

Reply:
only in places where done() is determined by the current state. This is not the case for the training_epoch_loop where we also check max_steps, which is checked against global counters. That is why I specifically pulled out the check for `num_training_batches_reached`.